### PR TITLE
fontloader update

### PR DESCRIPTION
**Pull request by phi-gamma**
No more Fontforge libraries ``=)``. This brings the basic functionality
for 2.7 with the goal of providing a release for Texlive 2016. Largely
untested except for generating the names index.

### CODE REVIEW

**Comment by eroux** [comment]
Wow, that looks like a great change! How does it work without fontforge? does it parse ttf in pure lua? Or use something like harfbuzz?

**Comment by phi-gamma** [comment]
···<date: 2016-04-17, Sunday>···<from: Elie Roux>···

> Wow, that looks like a great change!

Indeed, I expect a lot of angry users reporting breakage ``;)``

>                                      How does it work without
> fontforge? does it parse ttf in pure lua? Or use something like
> harfbuzz?

The former. Fonts are now read using the Lua loader that has been
in the making for about a year now. It features routines for
parsing the SFNT container and processing the outlines:

    https://git.phi-gamma.net/ConTeXt/tree/tex/context/base/mkiv/font-dsp.lua
    https://git.phi-gamma.net/ConTeXt/tree/tex/context/base/mkiv/font-cff.lua
    https://git.phi-gamma.net/ConTeXt/tree/tex/context/base/mkiv/font-ttf.lua

Basically, the Fontforge code in Luatex just became redundant.

With the new loader might be a tad slower to process an entire
font initially, but once it’s been cached everything else will
perform better. Font names indexing, btw., became much faster
since it’s no longer necessary to parse the OT tables of each
font as with FF. Not to mention memory usage drops dramatically.
There’s some stats at the top of ``luaotfload-database.lua``:

    https://github.com/phi-gamma/luaotfload/blob/master/src/luaotfload-database.lua

(Too bad we can’t use Valgrind anymore for debugging. That’s one
of the few downsides I can think of.)
